File: sdk/typescript/src/middleware.ts
import type { Context, Next } from 'hono';
import type { AuthAgentConfig, AuthUser } from './types.js';
import { AuthAgentClient } from './client.js';
import { AuthenticationError, AuthorizationError } from './types.js';

/**
 * Extract Bearer token from Authorization header
 */
function extractBearerToken(authHeader: string | undefined): string | null {
  if (!authHeader) return null;

  const match = authHeader.match(/^Bearer\s+(.+)$/i);
  return match ? match[1] : null;
}

/**
 * Hono middleware for Auth-Agent MCP
 *
 * @example
 * ```typescript
 * import { Hono } from 'hono';
 * import { authAgentMiddleware } from '@auth-agent/mcp-sdk';
 *
 * const app = new Hono();
 *
 * app.use('*', authAgentMiddleware({
 *   serverId: 'srv_abc123',
 *   apiKey: 'sk_xyz789',
 *   requiredScopes: ['files:read']
 * }));
 *
 * app.get('/files', (c) => {
 *   const user = c.get('user');
 *   return c.json({ user });
 * });
 * ```
 */
export function authAgentMiddleware(config: AuthAgentConfig) {
  const client = new AuthAgentClient(config);

  return async (c: Context, next: Next) => {
    try {
      // Extract token from Authorization header
      const authHeader = c.req.header('Authorization');
      const token = extractBearerToken(authHeader);

      if (!token) {
        return c.json(
          {
            error: 'unauthorized',
            error_description: 'Missing or invalid Authorization header',
          },
          401,
          {
            'WWW-Authenticate': 'Bearer realm="MCP Server"',
          }
        );
      }

      // Validate token
      const user = await client.validateToken(token);

      // Check scopes
      if (!client.hasRequiredScopes(user)) {
        return c.json(
          {
            error: 'insufficient_scope',
            error_description: `Required scopes: ${config.requiredScopes?.join(', ')}`,
            required_scopes: config.requiredScopes,
          },
          403
        );
      }

      // Attach user to context
      c.set('user', user);

      await next();
    } catch (error) {
      if (error instanceof AuthenticationError) {
        return c.json(
          {
            error: error.code || 'invalid_token',
            error_description: error.message,
          },
          401,
          {
            'WWW-Authenticate': 'Bearer realm="MCP Server", error="invalid_token"',
          }
        );
      }

      if (error instanceof AuthorizationError) {
        return c.json(
          {
            error: 'insufficient_scope',
            error_description: error.message,
            required_scopes: error.requiredScopes,
          },
          403
        );
      }

      // Unknown error
      console.error('Auth-Agent middleware error:', error);
      return c.json(
        {
          error: 'server_error',
          error_description: 'Authentication failed',
        },
        500
      );
    }
  };
}

/**
 * Extend Hono context with user type
 */
declare module 'hono' {
  interface ContextVariableMap {
    user: AuthUser;
  }
}

File: sdk/python/auth_agent_mcp/middleware.py
"""
FastAPI middleware for Auth-Agent MCP authentication
"""

import httpx
from typing import List, Optional
from fastapi import Request, HTTPException
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware


class AuthAgentMiddleware(BaseHTTPMiddleware):
    """
    Middleware to validate Auth-Agent access tokens for MCP servers.

    Example:
        from fastapi import FastAPI
        from auth_agent_mcp import AuthAgentMiddleware

        app = FastAPI()
        app.add_middleware(
            AuthAgentMiddleware,
            auth_server="https://mcp.auth-agent.com",
            server_id="srv_abc123",
            api_key="sk_xyz789",
            required_scopes=["files:read"]
        )
    """

    def __init__(
        self,
        app,
        auth_server: str = "https://mcp.auth-agent.com",
        server_id: Optional[str] = None,
        api_key: Optional[str] = None,
        required_scopes: Optional[List[str]] = None,
        public_paths: Optional[List[str]] = None,
    ):
        super().__init__(app)
        self.auth_server = auth_server.rstrip('/')
        self.server_id = server_id
        self.api_key = api_key
        self.required_scopes = required_scopes or []
        self.public_paths = public_paths or ['/health', '/']
        self.introspect_url = f"{self.auth_server}/introspect"

    async def dispatch(self, request: Request, call_next):
        # Skip auth for public endpoints
        if request.url.path in self.public_paths:
            return await call_next(request)

        # Extract token from Authorization header
        auth_header = request.headers.get("Authorization", "")
        if not auth_header.startswith("Bearer "):
            return self._unauthorized_response(request)

        token = auth_header[7:]  # Remove "Bearer "

        # Validate token with Auth-Agent
        try:
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    self.introspect_url,
                    headers={"Authorization": f"Bearer {self.api_key}"},
                    json={"token": token},
                    timeout=5.0,
                )
        except httpx.RequestError as e:
            raise HTTPException(
                status_code=503,
                detail=f"Auth server unavailable: {str(e)}"
            )

        if response.status_code != 200:
            return self._unauthorized_response(request)

        token_data = response.json()

        if not token_data.get("active"):
            return self._unauthorized_response(request)

        # Check scopes
        granted_scopes = token_data.get("scope", "").split()
        missing_scopes = [s for s in self.required_scopes if s not in granted_scopes]

        if missing_scopes:
            return self._forbidden_response(request, self.required_scopes)

        # Inject user context into request state
        request.state.user_email = token_data.get("sub")
        request.state.scopes = granted_scopes
        request.state.client_id = token_data.get("client_id")
        request.state.audience = token_data.get("aud")

        response = await call_next(request)
        return response

    def _unauthorized_response(self, request: Request) -> JSONResponse:
        """Return 401 Unauthorized with WWW-Authenticate header."""
        www_authenticate = (
            f'Bearer realm="{self.server_id or "mcp-server"}"'
        )

        if self.server_id:
            www_authenticate += (
                f', resource_metadata="{self.auth_server}/.well-known/'
                f'oauth-protected-resource/{self.server_id}"'
            )

        return JSONResponse(
            status_code=401,
            content={"error": "unauthorized"},
            headers={"WWW-Authenticate": www_authenticate}
        )

    def _forbidden_response(
        self,
        request: Request,
        required_scopes: List[str]
    ) -> JSONResponse:
        """Return 403 Forbidden with insufficient_scope error."""
        www_authenticate = (
            f'Bearer error="insufficient_scope", '
            f'scope="{" ".join(required_scopes)}"'
        )

        if self.server_id:
            www_authenticate += (
                f', resource_metadata="{self.auth_server}/.well-known/'
                f'oauth-protected-resource/{self.server_id}"'
            )

        return JSONResponse(
            status_code=403,
            content={
                "error": "insufficient_scope",
                "required_scopes": required_scopes
            },
            headers={"WWW-Authenticate": www_authenticate}
        )
